### PR TITLE
Feat: GLM-4.6 supports shared experts fusion

### PR DESCRIPTION
**Pull request by UranusSeven**
Hi from [novita.ai](https://novita.ai/) team 👋

## Motivation
Fuse shared experts with routed experts for better performance.

## Modifications
The changes involve modifying the model initialization to treat the shared expert as a regular expert, updating weight loading to remap shared expert weights.

## Accuracy Tests
```
python3 -m sglang.test.run_eval --port 30000 --eval-name mmlu --num-examples 200
ChatCompletionSampler initialized with self.system_message=None self.temperature=0.0 self.max_tokens=2048 self.reasoning_effort=None self.extra_body={}
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:01<00:00,  3.26it/s]
Writing report to /tmp/mmlu__models_glm4.6-fp8_snapshots_c064d336a8d0b0f59071f77eafdcdfca40f4b54c.html
{'other': np.float64(0.8936170212765957), 'other:std': np.float64(0.3083271648125413), 'score:std': np.float64(0.4208325082500163), 'stem': np.float64(0.7142857142857143), 'stem:std': np.float64(0.45175395145262554), 'humanities': np.float64(0.6935483870967742), 'humanities:std': np.float64(0.4610195460631114), 'social_sciences': np.float64(0.7959183673469388), 'social_sciences:std': np.float64(0.403028931798602), 'score': np.float64(0.77)}
Writing results to /tmp/mmlu__models_glm4.6-fp8_snapshots_c064d336a8d0b0f59071f77eafdcdfca40f4b54c.json
Total latency: 61.339 s
Score: 0.770
```

## Benchmarking and Profiling
| Metric | Before (Avg) | After (Avg) | Improvement |
|--------|--------------|-------------|-------------|
| **Request throughput (req/s)** | 3.845 | 3.845 | ~0% |
| **Input token throughput (tok/s)** | 15,922.76 | 15,939.07 | +0.10% |
| **Output token throughput (tok/s)** | 953.16 | 954.20 | +0.11% |
| **Total token throughput (tok/s)** | 16,875.92 | 16,894.26 | +0.11% |
| **Concurrency** | 81.47 | 64.49 | ↓ 20.8% |
| **Mean TTFT (ms)** | 824.10 | 628.65 | ↓ 23.7% |
| **Median TTFT (ms)** | 731.83 | 547.60 | ↓ 25.2% |
| **P99 TTFT (ms)** | 1997.92 | 1508.66 | ↓ 24.5% |
| **Mean TPOT (ms)** | 86.28 | 67.77 | ↓ 21.5% |
| **Median TPOT (ms)** | 86.40 | 68.20 | ↓ 21.0% |
| **P99 TPOT (ms)** | 217.53 | 145.96 | ↓ 32.9% |
| **Mean E2E latency (ms)** | 21190.94 | 16755.86 | ↓ 20.9% |
| **Median E2E latency (ms)** | 19153.18 | 14654.09 | ↓ 23.5% |
| **Mean ITL (ms)** | 82.48 | 65.32 | ↓ 20.8% |
| **P95 ITL (ms)** | 285.05 | 223.27 | ↓ 21.7% |
| **P99 ITL (ms)** | 1319.61 | 897.92 | ↓ 31.9% |
| **Max ITL (ms)** | 4457.03 | 2927.12 | ↓ 34.3% |

With shared expert fusion:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.0
Max request concurrency:                 128
Successful requests:                     500
Benchmark duration (s):                  130.06
Total input tokens:                      2070768
Total input text tokens:                 2070768
Total input vision tokens:               0
Total generated tokens:                  123959
Total generated tokens (retokenized):    123379
Request throughput (req/s):              3.84
Input token throughput (tok/s):          15921.57
Output token throughput (tok/s):         953.09
Total token throughput (tok/s):          16874.65
Concurrency:                             65.28
Accept length:                           1.97
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   16979.55
Median E2E Latency (ms):                 14778.42
---------------Time to First Token----------------
Mean TTFT (ms):                          638.74
Median TTFT (ms):                        555.73
P99 TTFT (ms):                           1570.34
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          68.80
Median TPOT (ms):                        68.77
P99 TPOT (ms):                           147.36
---------------Inter-Token Latency----------------
Mean ITL (ms):                           66.18
Median ITL (ms):                         21.34
P95 ITL (ms):                            223.45
P99 ITL (ms):                            875.24
Max ITL (ms):                            2923.89
==================================================

============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.0
Max request concurrency:                 128
Successful requests:                     500
Benchmark duration (s):                  129.76
Total input tokens:                      2070768
Total input text tokens:                 2070768
Total input vision tokens:               0
Total generated tokens:                  123959
Total generated tokens (retokenized):    123396
Request throughput (req/s):              3.85
Input token throughput (tok/s):          15958.57
Output token throughput (tok/s):         955.30
Total token throughput (tok/s):          16913.87
Concurrency:                             63.70
Accept length:                           1.97
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   16532.17
Median E2E Latency (ms):                 14529.75
---------------Time to First Token----------------
Mean TTFT (ms):                          618.56
Median TTFT (ms):                        539.46
P99 TTFT (ms):                           1446.97
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          66.73
Median TPOT (ms):                        67.63
P99 TPOT (ms):                           144.55
---------------Inter-Token Latency----------------
Mean ITL (ms):                           64.45
Median ITL (ms):                         20.26
P95 ITL (ms):                            223.09
P99 ITL (ms):                            920.60
Max ITL (ms):                            2930.35
==================================================
```

Without:
```
============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.0
Max request concurrency:                 128
Successful requests:                     500
Benchmark duration (s):                  130.15
Total input tokens:                      2070768
Total input text tokens:                 2070768
Total input vision tokens:               0
Total generated tokens:                  123959
Total generated tokens (retokenized):    123511
Request throughput (req/s):              3.84
Input token throughput (tok/s):          15910.88
Output token throughput (tok/s):         952.45
Total token throughput (tok/s):          16863.33
Concurrency:                             82.89
Accept length:                           1.97
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   21575.59
Median E2E Latency (ms):                 19498.84
---------------Time to First Token----------------
Mean TTFT (ms):                          830.75
Median TTFT (ms):                        746.95
P99 TTFT (ms):                           1920.41
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          87.94
Median TPOT (ms):                        87.86
P99 TPOT (ms):                           235.35
---------------Inter-Token Latency----------------
Mean ITL (ms):                           84.01
Median ITL (ms):                         20.68
P95 ITL (ms):                            278.54
P99 ITL (ms):                            1333.52
Max ITL (ms):                            4462.01
==================================================

============ Serving Benchmark Result ============
Backend:                                 sglang
Traffic request rate:                    4.0
Max request concurrency:                 128
Successful requests:                     500
Benchmark duration (s):                  129.95
Total input tokens:                      2070768
Total input text tokens:                 2070768
Total input vision tokens:               0
Total generated tokens:                  123959
Total generated tokens (retokenized):    123553
Request throughput (req/s):              3.85
Input token throughput (tok/s):          15934.63
Output token throughput (tok/s):         953.87
Total token throughput (tok/s):          16888.50
Concurrency:                             80.05
Accept length:                           1.97
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   20806.28
Median E2E Latency (ms):                 18807.53
---------------Time to First Token----------------
Mean TTFT (ms):                          817.45
Median TTFT (ms):                        716.70
P99 TTFT (ms):                           2075.43
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          84.61
Median TPOT (ms):                        84.94
P99 TPOT (ms):                           199.71
---------------Inter-Token Latency----------------
Mean ITL (ms):                           80.95
Median ITL (ms):                         20.34
P95 ITL (ms):                            291.56
P99 ITL (ms):                            1305.70
Max ITL (ms):                            4452.04
==================================================
```

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.ai/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/models/glm4_moe.py [hunk 1]
```diff
@@ -352,8 +353,14 @@ def __init__(
...
         self.top_k = config.num_experts_per_tok
         self.tp_size = get_tensor_model_parallel_world_size()
+        self.moe_ep_size = get_moe_expert_parallel_world_size()
         self.routed_scaling_factor = config.routed_scaling_factor
         self.n_shared_experts = config.n_shared_experts
+        self.num_fused_shared_experts = (
+            0
+            if get_global_server_args().disable_shared_experts_fusion
+            else config.n_shared_experts
+        )
```
[the remark this answers is not in the document]
Comment (UranusSeven) [reply python/sglang/srt/models/glm4_moe.py:363]: Crash is preferred if the config is incorrect.

**Review by ShangmingCai — COMMENTED** [review]
Looks great.

CC: @zRzRzRzRzRzRzR can you help review this PR?

**Comment by Kevin-XiongC** [comment]
``` bash
python3 -m sglang.test.run_eval --port 30000 --eval-name mmlu --num-examples 200
ChatCompletionSampler initialized with self.system_message=None self.temperature=0.0 self.max_tokens=2048 self.reasoning_effort=None self.extra_body={}
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:01<00:00,  3.26it/s]
Writing report to /tmp/mmlu__models_glm4.6-fp8_snapshots_c064d336a8d0b0f59071f77eafdcdfca40f4b54c.html
{'other': np.float64(0.8936170212765957), 'other:std': np.float64(0.3083271648125413), 'score:std': np.float64(0.4208325082500163), 'stem': np.float64(0.7142857142857143), 'stem:std': np.float64(0.45175395145262554), 'humanities': np.float64(0.6935483870967742), 'humanities:std': np.float64(0.4610195460631114), 'social_sciences': np.float64(0.7959183673469388), 'social_sciences:std': np.float64(0.403028931798602), 'score': np.float64(0.77)}
Writing results to /tmp/mmlu__models_glm4.6-fp8_snapshots_c064d336a8d0b0f59071f77eafdcdfca40f4b54c.json
Total latency: 61.339 s
Score: 0.770
```
The Accucracy test is attached above.

File: python/sglang/srt/models/glm4_moe.py [hunk 2]
```diff
@@ -1011,6 +1027,36 @@ def __init__(
     def get_input_embeddings(self) -> nn.Embedding:
         return self.model.embed_tokens
 
+    def determine_num_fused_shared_experts(self):
+        self.num_fused_shared_experts = 0
```
Comment (BBuf) [inline python/sglang/srt/models/glm4_moe.py]: `self.num_fused_shared_experts = 0` can be moved down.

File: python/sglang/srt/models/glm4_moe.py [hunk 3]
```diff
@@ -1065,11 +1111,15 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]], is_nextn=Fal
             ("gate_up_proj", "up_proj", 1),
         ]
 
+        num_fused_shared_experts = getattr(self, "num_fused_shared_experts", 0)
```
Comment (BBuf) [inline python/sglang/srt/models/glm4_moe.py]: Don't add logs inside forward; put them in `__init__` instead.

File: python/sglang/srt/models/glm4_moe.py [hunk 4]
```diff
@@ -1086,6 +1136,12 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]], is_nextn=Fal
         for name, loaded_weight in weights:
             weight_names.append(name)
 
+            if num_fused_shared_experts > 0 and "mlp.shared_experts" in name:
```
Comment (BBuf) [inline python/sglang/srt/models/glm4_moe.py]:
```shell
if num_fused_shared_experts > 0 and "mlp.shared_experts" in name:
    # Map shared expert weights to the last expert slot
    # For GLM4, we only support 1 shared expert
    assert self.config.n_shared_experts == 1, \
        f"Expected 1 shared expert, got {self.config.n_shared_experts}"
    
    # Shared expert becomes expert ID = n_routed_experts
    name = name.replace(
        "mlp.shared_experts",
        f"mlp.experts.{self.config.n_routed_experts}",
    )
```

Comment (jimmy-evo) [reply python/sglang/srt/models/glm4_moe.py]: > assert self.config.n_shared_experts == 1

this assertion is in `determine_num_fused_shared_experts`

File: python/sglang/srt/models/glm4_moe.py (hunk 2 above)
Comment (jimmy-evo) [reply python/sglang/srt/models/glm4_moe.py]: set in `__init__` as deepseek_v2 does so

**Comment by jimmy-evo** [comment]
@BBuf
thx for reviewing. refactored and applied your suggestion

**Review by BBuf — APPROVED** [review]
LGTM.
